### PR TITLE
Smithy Document Shape Support

### PR DESCRIPTION
**Pull request by skmcgrail**
Adds support for Smithy Document shapes and supporting types for protocols to implement support.

AWS Protocol Support: https://github.com/aws/aws-sdk-go-v2/pull/1320

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: document/json/encoder.go [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+	// Handle both pointers and interface conversion into types
+	rv = serde.ValueElem(rv)
+
+	switch rv.Kind() {
+	case reflect.Invalid:
+		if tag.OmitEmpty {
+			return nil
+		}
+		vp.GetValue().Null()
+		return nil
```
Comment (jasdel) [inline document/json/encoder.go:84]: would this case be hit if the `IsZeroValue` check above is triggered?

File: document/json/encoder.go [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+
+	default:
+		return e.encodeScalar(vp, rv)
+	}
+}
+
+func (e *Encoder) encodeZeroValue(vp valueProvider, rv reflect.Value) error {
+	switch rv.Kind() {
+	case reflect.Invalid:
+		vp.GetValue().Null()
```
Comment (jasdel) [inline document/json/encoder.go:107]: I think this duplicates the `Invalid` case in `encode`. the case in `encode` probably isn't needed?

File: document/json/encoder_test.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	})
+}
+
+func TestNewEncoderUnsupportedTypes(t *testing.T) {
+	type customTime time.Time
+
+	cases := []interface{}{
+		time.Now().UTC(),
+		customTime(time.Now().UTC()),
+	}
```
Comment (jasdel) [inline document/json/encoder_test.go:48]: Should this include cases for functions and channels?

File: document/json/json.go [hunk 4]
```diff
@@ -0,0 +1,25 @@
+package json
+
+func NewEncoder(optFns ...func(options *EncoderOptions)) *Encoder {
```
Comment (jasdel) [inline document/json/json.go:3]: nit documentation

File: document/json/decoder_test.go [hunk 5]
```diff
@@ -0,0 +1,630 @@
...
+		t.Fatalf("expect error, got nil")
+	}
+}
+
+func TestNewDecoderUnsupportedTypes(t *testing.T) {
+	cases := []struct {
+		input             []byte
+		value             interface{}
+		disableJSONNumber bool
+	}{
```
Comment (jasdel) [inline document/json/decoder_test.go:566]: Would be helpful to also have a test case that the `Decoder` cannot unmarshal into a non-empty interface.

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ShapeValueGenerator.java [hunk 6]
```diff
@@ -395,6 +407,107 @@ public Config build() {
...
+        public Void nullNode(NullNode node) {
+            writer.writeInline("nil");
+            return null;
+        }
+
+        @Override
+        public Void numberNode(NumberNode node) {
+            if (node.isNaturalNumber()) {
+                writer.writeInline("$L", node.getValue());
+            } else {
```
[the remark this answers is not in the document]
Comment (skmcgrail) [reply codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ShapeValueGenerator.java:448]: The else branch here covers Floats in the case that it was a BigDecimal type  according to codegen. Will need to investigate adding big.Int for the natural number case here.

File: document/document.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+var _ noSerde = (*NoSerde)(nil)
+
+// IsNoSerde returns whether the given type implements the no smithy document serde interface.
+func IsNoSerde(x interface{}) bool {
+	_, ok := x.(noSerde)
+	return ok
+}
+
+// Number is a arbitrary precision numerical value
+type Number string
```
[the remark this answers is not in the document]
Comment (skmcgrail) [reply document/document.go:34]: Yup I can add helpers for converting to those types.

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ShapeValueGenerator.java [hunk 8]
```diff
@@ -395,6 +407,107 @@ public Config build() {
...
+                } else if (value instanceof Double) {
+                    writer.writeInline("float64($L)", value.doubleValue(), value);
+                } else {
+                    writer.addUseImports(SmithyGoDependency.BIG);
+                    writer.writeInline("func () *big.Float {\n"
+                            + "\tf, ok := (&big.Float{}).SetString($S)\n"
+                            + "\tif !ok { panic(\"failed to parse string to float: \" + $S) }\n"
+                            + "\treturn f\n"
+                            + "}()", value, value);
+                }
```
[the remark this answers is not in the document]
Comment (skmcgrail) [reply codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ShapeValueGenerator.java:461]: They do not currently, can raise that with the Smithy team, overall I think document types is lacking in sufficient protocol tests.

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ProtocolDocumentGenerator.java [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+            }).write("");
+
+            writer.write("var _ $T = ($P)(nil)",
+                    serviceSmithyDocumentInterface, internalDocumentInterface);
+            writer.write("var _ $T = ($P)(nil)",
+                    smithyDocumentMarshaler, internalDocumentInterface);
+            writer.write("var _ $T = ($P)(nil)",
+                    smithyDocumentUnmarshaler,
+                    internalDocumentInterface);
+            writer.write("");
```
[the remark this answers is not in the document]
Comment (skmcgrail) [reply codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ProtocolDocumentGenerator.java:159]: Technically no harm in it being here but can move to a separate file.

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/ProtocolGenerator.java [hunk 10]
```diff
@@ -252,90 +258,171 @@ static String getDeserializeMiddlewareName(ShapeId operationShapeId, ServiceShap
...
+     * @param context   the generation context
      * @param operation the operation shape to retrieve errors for
      * @return map of error names to {@link ShapeId}
      */
     default Map<String, ShapeId> getOperationErrors(GenerationContext context, OperationShape operation) {
         return HttpProtocolGeneratorUtils.getOperationErrors(context, operation);
     }
 
+    default void generateProtocolDocumentMarshalerUnmarshalDocument(GenerationContext context) {
+        throw new CodegenException("document types not implemented for protocol");
```
Comment (mtdowling) [inline codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/ProtocolGenerator.java:270]: Can these all say which protocol?

Comment (mtdowling) [reply codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/ProtocolGenerator.java:270]: And can they get docs?

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/ProtocolGenerator.java [hunk 11]
```diff
@@ -252,90 +258,171 @@ static String getDeserializeMiddlewareName(ShapeId operationShapeId, ServiceShap
...
+
+    default void generateProtocolDocumentUnmarshalerUnmarshalDocument(GenerationContext context) {
+        throw new CodegenException("document types not implemented for protocol");
+    }
+
+    default void generateProtocolDocumentUnmarshalerMarshalDocument(GenerationContext context) {
+        throw new CodegenException("document types not implemented for protocol");
+    }
+
+    default void generateNewDocumentMarshaler(GenerationContext context, Symbol marshalerSymbol) {
```
Comment (mtdowling) [inline codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/ProtocolGenerator.java:285]: How are these different than the above? I guess docs would help here too a lot.

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/ProtocolGenerator.java [hunk 12]
```diff
@@ -252,90 +258,171 @@ static String getDeserializeMiddlewareName(ShapeId operationShapeId, ServiceShap
...
+        private final List<GoIntegration> integrations;
+        private final String protocolName;
+        private final GoDelegator delegator;
+
+        private GenerationContext(Builder builder) {
+            this.settings = SmithyBuilder.requiredState("settings", builder.settings);
+            this.model = SmithyBuilder.requiredState("model", builder.model);
+            this.service = SmithyBuilder.requiredState("service", builder.service);
+            this.symbolProvider = SmithyBuilder.requiredState("symbolProvider", builder.symbolProvider);
+            this.writer = builder.writer;
```
Comment (mtdowling) [inline codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/ProtocolGenerator.java:317]: Why is the writer optional?

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ProtocolDocumentGenerator.java [hunk 13]
```diff
@@ -0,0 +1,373 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (mtdowling) [inline codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ProtocolDocumentGenerator.java:2]: nit: 2021 on all the files

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ProtocolDocumentGenerator.java [hunk 14]
```diff
@@ -0,0 +1,373 @@
...
+import software.amazon.smithy.codegen.core.Symbol;
+import software.amazon.smithy.go.codegen.integration.ProtocolGenerator;
+import software.amazon.smithy.go.codegen.integration.ProtocolGenerator.GenerationContext;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.shapes.DocumentShape;
+
+/**
+ * Generates the Smithy document type for the service client.
+ */
+public final class ProtocolDocumentGenerator {
```
Comment (mtdowling) [inline codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ProtocolDocumentGenerator.java:31]: Does this class need to be public?

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ProtocolDocumentGenerator.java [hunk 15]
```diff
@@ -0,0 +1,373 @@
...
+import java.util.function.Consumer;
+
+import software.amazon.smithy.codegen.core.Symbol;
+import software.amazon.smithy.go.codegen.integration.ProtocolGenerator;
+import software.amazon.smithy.go.codegen.integration.ProtocolGenerator.GenerationContext;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.shapes.DocumentShape;
+
+/**
+ * Generates the Smithy document type for the service client.
```
Comment (mtdowling) [inline codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ProtocolDocumentGenerator.java:29]: Can you give more info on what this generates? Do you generate a document type per type defined in the model? Do you generate a single document type used by every document type, and this mostly just generates the serde code specific to the protocol?

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ProtocolDocumentGenerator.java [hunk 16]
```diff
@@ -0,0 +1,373 @@
...
+    public boolean hasDocumentShapes() {
+        return getDocumentShapes().size() > 0;
+    }
+
+    /**
+     * Add a document shape to the generator.
+     *
+     * @param shape the document shape to add.
+     */
+    public void addDocumentShape(DocumentShape shape) {
```
Comment (mtdowling) [inline codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ProtocolDocumentGenerator.java:83]: Isn't the set of document shapes knowable in the constructor? Something like:

```
context.getModel.getDocumentShapes();
```

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ProtocolDocumentGenerator.java [hunk 17]
```diff
@@ -0,0 +1,373 @@
...
+        delegator.useShapeWriter(settings.getService(model), writer -> {
+            writer.write("type $L = $T", NO_DOCUMENT_SERDE_TYPE_NAME, noSerde);
+        });
+
+        delegator.useFileWriter("./types/types.go", settings.getModuleName() + "/types", writer -> {
+            writer.write("type $L = $T", NO_DOCUMENT_SERDE_TYPE_NAME, noSerde);
+        });
+    }
+
+    private void generateInternalDocumentInterface() {
```
Comment (mtdowling) [inline codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ProtocolDocumentGenerator.java:128]: Does each service have its own document interface?

File: document/json/encoder.go [hunk 18]
```diff
@@ -0,0 +1,318 @@
...
+	"reflect"
+
+	"github.com/aws/smithy-go/document"
+	"github.com/aws/smithy-go/document/internal/serde"
+	smithyjson "github.com/aws/smithy-go/encoding/json"
+)
+
+type EncoderOptions struct{}
+
+type Encoder struct {
```
Comment (JordonPhillips) [inline document/json/encoder.go:15]: Why is this necessary? You appear to be using the standard library json marshaler anyway.

File: document/json/decoder.go [hunk 19]
```diff
@@ -0,0 +1,333 @@
...
+	"math/big"
+	"reflect"
+
+	"github.com/aws/smithy-go/document"
+	"github.com/aws/smithy-go/document/internal/serde"
+)
+
+type DecoderOptions struct{}
+
+type Decoder struct {
```
Comment (JordonPhillips) [inline document/json/decoder.go:15]: Why is this needed as opposed to the default json unmarshaler?

**Review by JordonPhillips — CHANGES_REQUESTED** [review]

File: document/internal/serde/field.go [hunk 20]
```diff
@@ -0,0 +1,292 @@
+package serde
```
[the remark this answers is not in the document]
Comment (skmcgrail) [reply document/internal/serde/field.go:1]: The only reasonable way I can think of would require us to just make this a publicly accessible Go package in smithy-go and we would have to maintain the public interfaces and forward compatability. It's certainly something we could do, and it would likely probably best done as a stand-alone module. Would we want to do that now or later as a separate task?

File: document/internal/serde/reflect.go [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+	BigFloat             reflect.Type
+	BigInt               reflect.Type
+	DocumentNumber       reflect.Type
+	MapStringToInterface reflect.Type
+	Time                 reflect.Type
+}{
+	BigFloat:             reflect.TypeOf((*big.Float)(nil)).Elem(),
+	BigInt:               reflect.TypeOf((*big.Int)(nil)).Elem(),
+	DocumentNumber:       reflect.TypeOf((*document.Number)(nil)).Elem(),
+	MapStringToInterface: reflect.TypeOf((map[string]interface{})(nil)),
```
[the remark this answers is not in the document]
Comment (skmcgrail) [reply document/internal/serde/reflect.go:20]: Not need in this instance given how the decoder / encoder is implemented.

File: document/internal/serde/serde.go [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+	if !v.IsValid() {
+		return nil
+	}
+	if v.Kind() == reflect.Ptr {
+		return PtrToValue(v.Interface())
+	}
+	return v.Interface()
+}
+
+func IsZeroValue(v reflect.Value) bool {
```
[the remark this answers is not in the document]
Comment (skmcgrail) [reply document/internal/serde/serde.go:64]: This is here to detect if a value is a zero-value, this doesn't prevent serialization of the zero value unless the user instructed the document marshaler via a struct tag like`document:",omitempty"`, where omitempty indicates that zero-values should not be serialized.

File: document/json/decoder.go (hunk 19 above)
Comment (skmcgrail) [reply document/json/decoder.go:15]: While document types are JSON-like we want to decouple them from the standard library's expectations and behaviors around the types it supports. (For example cases where Smithy has chosen not to support like Time serialization and blob serialization). By explicitly defining our own Encoder/Decoder interface and struct field tags we will be to evolve the interfaces further when modeled document types are added at a later date. Additionally, this enables us to have tighter control over type usage, such as being able to block usage of generated Smithy types as their document representation could change at a later date when modeled type support is added.

File: document/json/decoder.go [hunk 23]
```diff
@@ -0,0 +1,333 @@
...
+	case reflect.Struct:
+		if rv.CanInterface() && document.IsNoSerde(rv.Interface()) {
+			return &document.UnmarshalTypeError{
+				Value: fmt.Sprintf("unsupported type"),
+				Type:  rv.Type(),
+			}
+		}
+	case reflect.Interface:
+		rv.Set(reflect.MakeMap(serde.ReflectTypeOf.MapStringToInterface))
+		rv = rv.Elem()
```
[the remark this answers is not in the document]
Comment (skmcgrail) [reply document/json/decoder.go:295]: Yes this should :)

File: document/json/encoder.go (hunk 18 above)
Comment (skmcgrail) [reply document/json/encoder.go:15]: Going to resolve this comment, see earlier response on the Decoder for the teams reasoning.

File: document/json/encoder.go (hunk 1 above)
Comment (skmcgrail) [reply document/json/encoder.go:84]: Since `serde.ValueElem(rv)` unwraps pointers and interfaces to retrieve the element the resulting value in `rv` could still be Invalid.

File: document/json/encoder.go (hunk 2 above)
Comment (skmcgrail) [reply document/json/encoder.go:107]: See my reply to your earlier comment with regards to how Invalid is valid in both these cases.

File: document/json/encoder_test.go (hunk 3 above)
Comment (skmcgrail) [reply document/json/encoder_test.go:48]: The current behavior for channels and functions is to completely skip them and not serialize them (with no error returned), which is the behavior of the `encoding/json` library. I feel like it makes sense to at least attempt to match that behavior for those types. Time is a bit of a special case since we want to avoid making any sort of opinionated stance on their handling at this time, whereas channels and functions would more then likely never be supported.

File: document/document.go [hunk 24]
```diff
@@ -0,0 +1,71 @@
+package document
+
+import (
+	"strconv"
+)
+
+type SmithyDocumentMarshaler interface {
```
[the remark this answers is not in the document]
Comment (skmcgrail) [reply document/document.go:7]: Yes, next revision will include documentation, just wanted to get this out for initial feedback on the overall implementation.

File: document/document.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (skmcgrail) [reply document/document.go:34]: This is used in the Encoder/Decoder. This is a similar concept to the standard library's json.Number in that a user can use this type to capture a number literal into a string and then use one of the function methods to coerce to your desired type (int64, float64, etc).

File: document/errors.go [hunk 25]
```diff
@@ -0,0 +1,74 @@
...
+// Error returns the string representation of the error satisfying the error
+// interface.
+func (e *UnmarshalError) Error() string {
+	return fmt.Sprintf("unmarshal failed, cannot unmarshal %q into %s, %v",
+		e.Value, e.Type.String(), e.Err)
+}
+
+// An InvalidMarshalError is an error type representing an error
+// occurring when marshaling a Go value type to an AttributeValue.
+type InvalidMarshalError struct {
```
[the remark this answers is not in the document]
Comment (skmcgrail) [reply document/errors.go:66]: There isn't any cases on the Marshal code path where we wrap and underlying error currently which is why this type doesn't exist specifically.

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/DocumentShapeDeserVisitor.java [hunk 26]
```diff
@@ -411,7 +411,7 @@ protected final void generateDeserFunction(
             BiConsumer<GenerationContext, Shape> functionBody
     ) {
         SymbolProvider symbolProvider = context.getSymbolProvider();
-        GoWriter writer = context.getWriter();
+        GoWriter writer = context.getWriter().get();
```
[the remark this answers is not in the document]
Comment (skmcgrail) [reply codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/DocumentShapeDeserVisitor.java:414]: The ProtocolGenerator context is kind of a bit of a grab bag currently, in some instances a writer is to be expected, in other instances the delegator that is stored on the context is used and no writer is expected. This was just to make that contact a bit clearer that it may not be there. 

This was just something that I noticed when I was refactoring context to make it immutable after construction.

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/ProtocolGenerator.java (hunk 12 above)
Comment (skmcgrail) [reply codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/ProtocolGenerator.java:317]: I refer to this in an earlier comment, but since the generation context is a bit of a grab bag, the writer is not required in all instances of the context's usage, thus is why it is optional.

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/ProtocolUtils.java [hunk 27]
```diff
@@ -210,4 +210,8 @@ public static void writeSerDelegateFunction(
 
         lambda.accept(acceptVar);
     }
+
+    public static void writeDocumentInterfaceCheck(GenerationContext context, GoWriter writer) {
```
[the remark this answers is not in the document]
Comment (skmcgrail) [reply codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/ProtocolUtils.java:214]: This is an older artifact and didn't get cleaned up, will remove in next revision.

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ProtocolDocumentGenerator.java (hunk 14 above)
Comment (skmcgrail) [reply codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ProtocolDocumentGenerator.java:31]: Yes, these exported constants are needed so that protocol implementation parts make symbol references to the types.

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ProtocolDocumentGenerator.java (hunk 15 above)
Comment (skmcgrail) [reply codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ProtocolDocumentGenerator.java:29]: Will add some clarifying code examples sections to this class documentation. To summarize what is being created here: Each service client generates its own document type interface definition. This ensures that document types that are sent or received from the service are bound to the client package and it's associated client protocol. This prevents the situation where a user recieves

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ProtocolDocumentGenerator.java [hunk 28]
```diff
@@ -0,0 +1,373 @@
...
+            writer.openBlock("func $L(v interface{}) $T {", "}", NEW_LAZY_DOCUMENT,
+                            getDocumentSymbol(DOCUMENT_INTERFACE_NAME), () -> {
+                                writer.write("return $T(v)",
+                                        getInternalDocumentSymbol(INTERNAL_NEW_DOCUMENT_MARSHALER_FUNC));
+                            })
+                    .write("");
+        });
+    }
+
+    private void generateNoSerdeType() {
```
[the remark this answers is not in the document]
Comment (skmcgrail) [reply codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ProtocolDocumentGenerator.java:115]: This type is actually needed in all service client structures types to avoid making backwards incompatible changes when modeled smithy document types are later added at a future date. Since Smithy makes no opinionated stance currently on things like JSON object keys should be formatted (lower-case, camelCase, UPPER CASE) etc, this NoSerde type is embedded in a generated smithy structure types from the model to prevent their usage as document types. That way if an existing structure is later annotated as modeled document type, we don't have a situation where we make a potential breaking change in the representation of the underlying type.  (object key casing being an example).

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ProtocolDocumentGenerator.java (hunk 17 above)
Comment (skmcgrail) [reply codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ProtocolDocumentGenerator.java:128]: explained in an earlier comment above, going to resolve this in favor of that thread.

**Comment by skmcgrail** [comment]
I believe all outstanding feedback has been capture in the `Implement Feedback` commit that I have force-pushed. Please review when you have the chance, where applicable I left some conversations unresolved to give you the opportunity to read the response, feel free to resolve if satisfied.

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ProtocolDocumentGenerator.java [hunk 29]
```diff
@@ -0,0 +1,520 @@
...
+            writer.write("type $L = $T", DOCUMENT_INTERFACE_NAME,
+                            getInternalDocumentSymbol(DOCUMENT_INTERFACE_NAME))
+                    .write("");
+
+            writer.writeDocs(String.format("You create document type using the %s function and passing it the Go "
+                    + "type to be marshaled and sent to the service. The document marshaler supports semantics similar "
+                    + "to the encoding/json Go standard library.", NEW_LAZY_DOCUMENT));
+            writer.writeDocs("");
+            writer.writeDocs("For more information see the accompanying package documentation and linked references.");
+            writer.openBlock("func $L(v interface{}) $T {", "}", NEW_LAZY_DOCUMENT,
```
Comment (jasdel) [inline codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ProtocolDocumentGenerator.java:154]: Should this document the struct tag to use for custom member names?

File: document/document.go [hunk 30]
```diff
@@ -0,0 +1,153 @@
...
+)
+
+// Marshaler is an interface for a type that marshals a document to its protocol-specific byte representation and
+// returns the resulting bytes. A non-nil error will be returned if an error is encountered during marshaling.
+//
+// Marshal supports basic scalars (int,uint,float,bool,string), big.Int, and big.Float, maps, slices, and structs.
+// Anonymous nested types are flattened based on Go anonymous type visibility.
+//
+// When defining struct types. the `document` struct tag can be used to control how the value will be
+// marshaled into the resulting protocol document.
```
Comment (jasdel) [inline document/document.go:16]: Maybe this is where the earlier comment I had about linking to struct tags should link to? What do you think about linking to the smithy-go documentation within the code's docstring?

File: document/internal/serde/field.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (jasdel) [reply document/internal/serde/field.go:1]: Makes sense. That sounds like something we could add later as some kind of type walker helper if we decided its valuable in a maintainable way. We probably don't want to do that here.

**Review by jasdel — APPROVED** [review]

**Review by JordonPhillips — APPROVED** [review]

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ProtocolDocumentGenerator.java (hunk 29 above)
Comment (skmcgrail) [reply codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ProtocolDocumentGenerator.java:154]: The doc.go for the clients `document` package has a `doc.go` from this template file https://github.com/aws/smithy-go/blob/a87ef15659d457c607421ee90635d35b2e6def4c/codegen/smithy-go-codegen/src/main/resources/software/amazon/smithy/go/codegen/document_doc.go.template which details more details and links to the smithy-go document package.

File: document/document.go (hunk 30 above)
Comment (skmcgrail) [reply document/document.go:16]: https://github.com/aws/smithy-go/blob/a87ef15659d457c607421ee90635d35b2e6def4c/codegen/smithy-go-codegen/src/main/resources/software/amazon/smithy/go/codegen/document_doc.go.template links to this.

**Comment by skmcgrail** [comment]
Rebased on `main`.
